hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -159,6 +161,151 @@ def trivial(self):
+    or ``ConjectureData`` object.
+
+    Pretends to be a list containing ``Block`` objects but actually only
+    contains their endpoints right up until the point where you want to
+    access the actual block, at which point it is constructed.
+
+    This is designed to be as space efficient as possible, so will at
+    various points silently transform its representation into one
+    that is better suited for the current access pattern.
+
+    In addition, it has a number of convenience methods for accessing
+    properties of the block object at index ``i`` that should generally
+    be preferred to using the Block objects directly, as it will not
+    have to allocate the actual object."""
+
+    __slots__ = ("__endpoints", "owner", "__blocks", "__count", "__sparse")
+
+    def __init__(self, owner):
+        self.owner = owner
+        self.__endpoints = IntList()
+        self.__blocks = {}
+        self.__count = 0
+        self.__sparse = True
+
+    def add_endpoint(self, n):
+        """Add n to the list of endpoints."""
+        assert isinstance(self.owner, ConjectureData)
+        self.__endpoints.append(n)
+
+    def transfer_ownership(self, new_owner):
+        """Used to move ``Blocks`` over to a ``ConjectureResult`` object
+        when that is read to be used and we no longer want to keep the
+        whole ``ConjectureData`` around."""
+        assert isinstance(new_owner, ConjectureResult)
+        self.owner = new_owner
+
+    def start(self, i):
+        """Equivalent to self[i].start."""
+        if i == 0:
+            return 0
+        else:
+            return self.end(i - 1)
+
+    def end(self, i):
+        """Equivalent to self[i].end."""
+        return self.__endpoints[i]
+
+    def bounds(self, i):
+        """Equivalent to self[i].bounds."""
+        return (self.start(i), self.end(i))
+
+    def all_bounds(self):
+        """Equivalent to [(b.start, b.end) for b in self]."""
+        result = []
+        prev = 0
+        for e in self.__endpoints:
+            result.append((prev, e))
+            prev = e
+        return result
+
+    def __len__(self):
+        return len(self.__endpoints)
+
+    def __known_block(self, i):
+        try:
+            return self.__blocks[i]
+        except (KeyError, IndexError):
+            return None
+
+    def __getitem__(self, i):
+        n = len(self)
+        if i < -n or i >= n:
+            raise IndexError("Index %d out of range [-%d, %d)" % (i, n, n))
+        if i < 0:
+            i += n
+        assert i >= 0
+        result = self.__known_block(i)
+        if result is not None:
+            return result
+
+        # We store the blocks as a sparse dict mapping indices to the
+        # actual result, but this isn't the best representation once we
+        # stop being sparse and want to use most of the blocks. Switch
+        # over to a list at that point.
+        if self.__sparse and len(self.__blocks) * 2 >= len(self):
```

Comment:
I can imagine a scenario where this gets tripped early during generation (e.g. 50% of 2 blocks is 1 block), and then we end up switching over to a list when it would be better to stick with the dictionary for a while.

I don’t know how realistic that is, and the penalty should only be a modest space increase, so I don’t know if it’s worth worrying about.

Reply:
Yeah I wondered about that. My inclination that it was probably not worrying about until we see it in the wild as a significant problem - the usage patterns that trigger it are likely to be not that common and the failure mode is just increased memory usage. Also until we actually *see* it happening it's hard to know what a good heuristic for avoiding it is. I could add a lower bound if you want (say don't trigger it until we have at least 5 elements?) but I thought it might be better to wait and see.

Reply:
I think it's OK to try as-is.

Reply:
IMO we shouldn't complicate our optimisations unless and until we actually observe a problem.

At which point porting to Rust might be a better choice anyway...

Reply:
> At which point porting to Rust might be a better choice anyway...

I have found myself "goddammit I wish I'd moved this code to Rust already" *so* many times in the last two weeks.